src/cliffordsynthesis/encoding/TwoQubitEncoder.cpp

```diff
@@ -0,0 +1,251 @@
+      "h_" + std::to_string(pos) + "_" + std::to_string(xorHelper.size());
+  DEBUG() << "Creating helper variable for RChange XOR " << hName;
+  const auto n = static_cast<std::int16_t>(S);
+  xorHelper.emplace_back(lb->makeVariable(hName, CType::BITVECTOR, n));
+  if (xorHelper.size() == 1) {
+    lb->assertFormula(xorHelper.back() == changes);
+  } else {
+    lb->assertFormula(xorHelper.back() ==
+                      (xorHelper[xorHelpers[pos].size() - 2] ^ changes));
+  }
+}
+
+void encoding::TwoQubitEncoder::assertTwoQubitGateConstraints(
+    const std::size_t pos) {
+  const auto& twoQubitGates = vars.gC[pos / 2];
+  for (std::size_t ctrl = 0U; ctrl < N; ++ctrl) {
+    for (std::size_t trgt = 0U; trgt < N; ++trgt) {
+      if (ctrl == trgt) {
+        const auto changes = createIdentityConstraintOnTQG(pos, ctrl);
+        lb->assertFormula(
+            LogicTerm::implies(twoQubitGates[ctrl][ctrl], changes));
+        splitXorR(tvars->singleQubitRChange(pos, ctrl, qc::OpType::None), pos);
+      } else {
+        const auto changes = createTwoQubitGateConstraint(pos, ctrl, trgt);
+        lb->assertFormula(
+            LogicTerm::implies(twoQubitGates[ctrl][trgt], changes));
+        DEBUG() << "Asserting CNOT on " << ctrl << " and " << trgt;
+      }
+    }
+  }
+}
+
+LogicTerm
+encoding::TwoQubitEncoder::createIdentityConstraintOnTQG(std::size_t pos,
+                                                         std::size_t ctrl) {
+  auto changes = tvars->x[pos + 1][ctrl] == tvars->x[pos][ctrl];
+  changes      = changes && (tvars->z[pos + 1][ctrl] ==
+                        tvars->z[pos][ctrl]); // && here is overloaded
+
+  return changes;
+}
+
+LogicTerm encoding::TwoQubitEncoder::createTwoQubitGateConstraint(
+    std::size_t pos, std::size_t ctrl, std::size_t trgt) {
+  auto changes              = LogicTerm(true);
+  const auto [xCtrl, xTrgt] = tvars->twoQubitXChange(pos, ctrl, trgt);
+  const auto [zCtrl, zTrgt] = tvars->twoQubitZChange(pos, ctrl, trgt);
+
+  changes = changes && (tvars->x[pos + 1][ctrl] == xCtrl);
+  changes = changes && (tvars->x[pos + 1][trgt] == xTrgt);
+  changes = changes && (tvars->z[pos + 1][ctrl] == zCtrl);
+  changes = changes && (tvars->z[pos + 1][trgt] == zTrgt);
+
+  const auto& newRChanges = LogicTerm::ite(
+      vars.gC[pos / 2][ctrl][trgt], tvars->twoQubitRChange(pos, ctrl, trgt),
+      LogicTerm(0, static_cast<std::int16_t>(S)));
+  splitXorR(newRChanges, pos);
+
+  return changes;
+}
+
+void TwoQubitEncoder::extractCircuitFromModel(Results& res, Model& model) {
+  std::size_t nSingleQubitGates = 0U;
+  std::size_t nTwoQubitGates    = 0U;
+
+  qc::QuantumComputation qc(N);
+  for (std::size_t t = 0; t < T / 2; ++t) {
+    extractSingleQubitGatesFromModel(t, model, qc, nSingleQubitGates);
+    extractTwoQubitGatesFromModel(t, model, qc, nTwoQubitGates);
+  }
+
+  res.setSingleQubitGates(nSingleQubitGates);
+  res.setTwoQubitGates(nTwoQubitGates);
+  res.setTQDepth(qc.getTQDepth());
```

Comment:
This should be for the other encodings as well so you can compare against them. Add this computation to `GateEncoder::extractCircuitFromModel`